### PR TITLE
Add --one-based to support number of jobs starting from 1:

### PR DESCRIPTION
**Pull request by xsawyerx**
As part of the conversation in GH#3125, this is adding another CLI option `--one-based` which should turn the number of jobs to begin from 1 instead of from 0.

It is implemented to create the least amount of disruption, both to the code and to the checks it does on jobs.

Because all of the checks assume that "jobs == 0" means that `-j` was not used, I cannot override those variables early on.

Instead, I'm setting a flag and using it later to determine how many jobs we actually had. If both `-j` and `--one-based` are used, we count to one additional jobs when looping over them.

This will set different values for `@OUTPUTS` variable, which at this point carries the jobs and does not need to know how many there were.

### CODE REVIEW

**Comment by xsawyerx** [comment]
The number of commits seems massive, but there was only one commit added. The others are based on #3125. If that one is merged, I will be happy to update this to only reflect that one commit of difference. That should also make merging easier. :)

**Comment by xsawyerx** [comment]
Is there a chance to provide an example of input and expected output? Clearly, I don't fully understand the logic that is going on here and I'd rather not make another bad attempt at this.

**Comment by kkm000** [comment]
Sure. Here's code right from a script I have run logs for, slightly edited:
```sh
nj=16
run.pl JOB=1:$nj $dst/log/generate_alignments.JOB.log \
     lattice-best-path --acoustic-scale=$acoustic_scale \
       "scp:utils/split_scp.pl -j $nj \$((JOB-1)) $dst/src-lats.scp |" \
       ark:/dev/null "ark:|gzip -c >$dst/ali.JOB.gz" || exit 1;
```
`JOB` is a variable. run.pl runs multiple copies of its command in parallel. For each command, `JOB` takes successive indices as specified, in this case 1,...,16. run.pl's second argument is where it saves its tun log (16 of them). So we have 16 of this log (for an arbitrary value of 7; please ignore 7 in '7way', it's a directory name). Look how JOB is substituted
```
$ head -4 exp/gsm1/chain/ali_train_g100_7way_hires/log/generate_alignments.7.log
# lattice-best-path --acoustic-scale=0.125 "scp:utils/split_scp.pl -j 16 $((7-1)) exp/gsm1/chain/ali_train_g100_7way_hires/src-lats.scp |" ark:/dev/null "ark:|gzip -c >exp/gsm1/chain/ali_train_g100_7way_hires/ali.7.gz"
# Started at Sun Mar 24 04:21:26 PDT 2019
#
lattice-best-path --acoustic-scale=0.125 'scp:utils/split_scp.pl -j 16 6 exp/gsm1/chain/ali_train_g100_7way_hires/src-lats.scp |' ark:/dev/null 'ark:|gzip -c >exp/gsm1/chain/ali_train_g100_7way_hires/ali.7.gz'
```
So run.pl command line chunk `split_scp.pl -j $nj \$((JOB-1))` became `split_scp.pl -j 16 $((7-1))` as run.pl saw it, then `split_scp.pl -j 16 6` as the command which it spawned, lattice-best-path, saw it and reported.

The key here is the rspecifier `'scp:utils/split_scp.pl -j 16 6 exp/gsm1/chain/ali_train_g100_7way_hires/src-lats.scp |`. split_scp is told to output only part 6 (counting from 0) of a 16-way fair, exhaustive and non-intersecting subset of lines from the file src-lats.scp. Of course, other copies of the script, run on different CPUs or machines, will each get their own number substituted for the `JOB`.

While we are always numbering jobs and parts and everything from 1, the script itself requires a 0-based number. That's kind of inconsistent.

You can test the script easily:
* Generate a test file with N unique lines, e. g., with seq. 
* Ask split_scp.pl to print every part K of N exhaustively (K=0..N-1 in 0-based case, K=1..N in 1-based case)
* Merge the parts, sort them, and make sure you got back the source file.

**Comment by kkm000** [comment]
Another invariant is split_scp.pl must not violate the ordering of lines property of the source file (if the source is sorted by some `<`, every subset is stable-sorted by the same `<`), but you are not touching this.

**Comment by kkm000** [comment]
Another invariant is split_scp.pl must not violate the ordering of lines property of the source file (if the source is sorted by some `<`, every subset is stable-sorted by the same `<`), but you are not touching this.

File: egs/wsj/s5/utils/split_scp.pl [hunk 1]
```diff
@@ -58,6 +58,10 @@
         $utt2spk_file = $1;
         shift;
     }
+    if ($ARGV[0] eq '--one-based') {
```
[the remark this answers is not in the document]
Comment (danpovey) [reply egs/wsj/s5/utils/split_scp.pl]: This comment by @kkm000 is still not addressed.  Fix is simple.

File: egs/wsj/s5/utils/split_scp.pl [hunk 2]
```diff
@@ -132,94 +160,129 @@
...
     # [just doing int() rounds down].
-    $n = 0;
-    for($scpidx = 0; $scpidx < @OUTPUTS; $scpidx++) {
-        $scpfile = $OUTPUTS[$scpidx];
-        open(O, ">$scpfile") || die "Opening output scp file $scpfile";
-        for($k = 0; $k < $linesperscp + ($scpidx < $remainder ? 1 : 0); $k++) {
-            print O $F[$n++];
+    my $n = 0;
+    for (my $scpidx = 0; $scpidx < @OUTPUTS; $scpidx++) {
+        my $scpfile = $OUTPUTS[$scpidx];
```
Comment (danpovey) [inline egs/wsj/s5/utils/split_scp.pl]: There seems to be quite a lot of reformatting done.  This makes it very hard for me to see what the actual changes are.  This is a script that we need to be careful about changing the behavior of.  The diff is way too large.

**Comment by danpovey** [comment]
@kkm000 do you feel like doing this?

**Comment by kkm000** [comment]
Sure thing, I like the 1-based feature. And the readability improvement is certainly a good thing in itself too, unless taken to the extreme as to go out-of-line with our other scripts.

**Comment by xsawyerx** [comment]
I'm sorry I didn't get around to making the changes. I'm reviewing them now.

**Comment by xsawyerx** [comment]
> Thanks for working on this, really!
> 
> If I were you, I'd just subtract 1 from `$job_id` right from the start. It seems a simpler solution:
> 
> ```perl
> my $job_offset=0;
> for ($x .... ) { # the argument parsing loop
>   if -j {  $job_id=...; $num_jobs=...;  die if $num_jobs <= 0 } # only validate num_jobs just yet
>   if --one-based { $job_offset=1 }
> }
> # After the loop, you can map $job_id into the existing range, and then do all the checks
> $job_id -= $job_offset  # This is same as the ugly $((JOB-1)) in scripts
> # Note $job_id is automatically negative if --one-based is provided w/o -j, so this one condition covers it all:
> die "..." unless $job_id >= 0 && ($job_id < $num_jobs || $num_jobs == 0)
> ```
> 
> (You could do fancier here and report _what_ exactly was wrong with arguments:
> 
>     1. `$num_jobs == 0 && $job_id < 0` =>> --one_based w/o -j
> 
>     2. `$job_id < 0 || $job_id + 1 >= $num_jobs` (assuming we did not die above) =>> invalid -j,
> 
> 
> but this is not really important. This script is quite low-level. Up to you, of course, but I probably would not.)
> 
> Nothing else seems to need to be changed. Does this make sense?

The reason I had not done this (just decremented `$job_id`) is because I wanted a less intrusive patch. There is a condition for `$num_jobs` and `@ARGV` and I didn't want to interfere with since I wasn't sure what it's supposed to do. Also, because no Getopt library is used, the `@ARGV` handling is fragile. (I presented in the other PR how it can be inconsistent.) I'm happy to adopt your proposal though.

Calling the variable `$job_offset` isn't accurate since it's not a general offset. It's a very specific offset - just by one. I'd rather keep it as `$one_based` because that's more accurate as a boolean of either starting with zero or starting with one.

I think there is some bug there with the offset. Because you consider it an offset instead of being "one-based," aren't you decrementing when it's supposed to be incremented? If, by default, it starts from zero and we want it to start from one, we would have to increment it, not decrement. Isn't that right?

**Pull request closed without merging**